lib/collection/src/update_handler.rs

```diff
@@ -541,95 +547,110 @@ impl UpdateHandler {
+                continue;
+            }
 
-                    // Determine optimization handle limit based on max handles we allow
-                    // Not related to the CPU budget, but a different limit for the maximum number
-                    // of concurrent concrete optimizations per shard as configured by the user in
-                    // the Qdrant configuration.
-                    // Skip if we reached limit, an ongoing optimization that finishes will trigger this loop again
-                    let limit = max_handles.saturating_sub(optimization_handles.lock().await.len());
-                    if limit == 0 {
-                        log::trace!(
-                            "Skipping optimization check, we reached optimization thread limit"
-                        );
-                        continue;
-                    }
+            if Self::try_recover(segments.clone(), wal.clone())
+                .await
+                .is_err()
+            {
+                continue;
+            }
 
-                    Self::process_optimization(
-                        optimizers.clone(),
-                        segments.clone(),
-                        optimization_handles.clone(),
-                        optimizers_log.clone(),
-                        total_optimized_points.clone(),
-                        &optimizer_cpu_budget,
+            // Continue if we have enough CPU budget available to start an optimization
+            // Otherwise skip now and start a task to trigger the optimizer again once CPU
+            // budget becomes available
+            let desired_cpus = num_rayon_threads(max_indexing_threads);
+            if !optimizer_cpu_budget.has_budget(desired_cpus) {
+                let trigger_active = cpu_available_trigger
+                    .as_ref()
+                    .map_or(false, |t| !t.is_finished());
+                if !trigger_active {
+                    cpu_available_trigger.replace(trigger_optimizers_on_cpu_budget(
+                        optimizer_cpu_budget.clone(),
+                        desired_cpus,
                         sender.clone(),
-                        limit,
-                    )
-                    .await;
+                    ));
                 }
+                continue;
             }
+
+            // Determine optimization handle limit based on max handles we allow
+            // Not related to the CPU budget, but a different limit for the maximum number
+            // of concurrent concrete optimizations per shard as configured by the user in
+            // the Qdrant configuration.
+            // Skip if we reached limit, an ongoing optimization that finishes will trigger this loop again
+            let limit = max_handles.saturating_sub(optimization_handles.lock().await.len());
+            if limit == 0 {
+                log::trace!("Skipping optimization check, we reached optimization thread limit");
```

Comment:
I never use `trace` because it is way too verbose.
Would it make sense to bump this one to `debug`?

Reply:
It can happen a lot and cause a lot of log spam in debug. A lot as in, with every update operation you send. That's why we demoted it a long time ago to trace.

It actually was useful in our benchmarking setup a few days ago where we changed logging to trace just to confirm some branches were hit.

Do you think setting it to debug would be worth the log spam?

Reply:
Let's keep it to `trace` if it is noisy :+1: